Vehicle editor + parts

dagriefaa opened the pull request:
adds:
- basic part definitions
- basic part prefabs
- vehicle editor
    - placing parts (with snapping, placing against part surfaces, and free placement)
    - deleting part trees
    - placing parts with symmetry (parent symmetry and root symmetry, root symmetry is working in some cases)

controls:
- [1] -> summon/cycle part
- [left mouse] -> place part
- [esc] -> cancel ghost
- [ x] -> destroy part tree

should be tested before merging

dagriefaa committed:
add basic orbital camera

dagriefaa committed:
add veeery basic part definitions

dagriefaa committed:
start on place/remove code

dagriefaa committed:
parts data + fix flaky raycasts + approaching a solution for deleting parts

dagriefaa committed:
part definition component

dagriefaa committed:
part deletion

dagriefaa committed:
really bad part placing + part tree deletion

dagriefaa committed:
cleanup + bounds calcs for ghost

dagriefaa committed:
add comments

dagriefaa committed:
when placing against something, don't clip ghost into it

dagriefaa committed:
add part snapping

dagriefaa commented:
if i saw a 5k delta on a pr at work i would die lol

Ragingram2 commented on Assets/Code/VehicleEditor/EditSystem.cs:
```diff
@@ -0,0 +1,310 @@
+
+    /// <returns>true if part found</returns>
+    bool GetPartFrom(RaycastInput raycast, out RaycastHit closestHit, out Entity part) {
+        if (!SystemAPI.GetSingleton<PhysicsWorldSingleton>().CastRay(raycast, out closestHit)) {
```
That SystemAPI.GetSingleton() call should probably be cached at startup. Potentially somewhere static so other systems can benefit from it. Same goes for any other call to SystemAPI.GetSingleton

dagriefaa replied:
1. in unity dots, data should not be stored in systems - entity queries, GetComponent, GetSingleton etc. in the update loop are the intended way of retrieving data (and as such are quite fast)
2. it should also be noted that ideally systems should inherit from `ISystem`, which is a struct that has no state of its own at all - `SystemBase` is only used here because it needs to interact with the input system

- https://learn.unity.com/tutorial/part-3-1-implementation-fundamentals?uv=2022.3&courseId=60132919edbc2a56f9d439c3&projectId=6013255bedbc2a2e590fbe60#639affd6edbc2a72e6eb288d

Ragingram2 replied:
ah, alright yeah that makes sense. The call is probably fast enough as it is.

Ragingram2 commented on Assets/Code/VehicleEditor/EditSystemDataAuthoring.cs:
```diff
@@ -0,0 +1,24 @@
+public class EditSystemDataAuthoring : MonoBehaviour {
+
+    [TextArea]
+    public string notes = "Did you make sure your new part \n- doesn't have nested children?\n- is 1x scale?";
```
Theres a pretty neat free and MIT licensed package called Naughty attributes (https://github.com/dbrizov/NaughtyAttributes) that make things like this easier to make. Also has a bunch of other really cool editor attribute stuff. Not something that should be added with this PR btw, just a suggestion

dagriefaa replied:
i agree but think that we should worry about niceties once we start on the full project

Ragingram2 reviewed (commented):
Not sure what sort of PR rules we wanna setup, So I'll leave my comments as just comments, Plus my knowledge of DOTS is still pretty lackluster. 

All in all this looks solid, and I've seen it working on Discord.

warior456 commented:
maybe some control documentation would be nice

UknownPI committed:
Basic (parent symmetry)

Just a basic symmetry working (only symmetry around the part you just clicked on)(so if you set symmetry x4 there will only be 4 part placed on the part you just clicked on)

UknownPI committed:
Half Finished Root Symmetry

(Only works when clicking on the main "branch" , I might take some time to finish this so I might as well put it on github for others to see)

UknownPI commented:
Ok so there are 2 mods of symmetry : One that's very simple and is only around the clicked part , and one more complicated root symmetry that is around all parts "similar" / "siblings" to the one clicked on (in the future this mods could change and maybe merge the two modes into one that allows you to choose how much parents back do you want to iterate on (where the parent mode from now is like iterating on one parent , and root is on all parents))
The root symmetry mode now works only when the part you clicked on is on the "Main Branch" (a branch where every part is the first of his "level"  / "Iteration")
![image](https://github.com/Project-Launch-Escape/unity-prototype/assets/169551292/c02dd42c-a47b-4ce3-9d16-4d0ff249abb6)
This image is an illustration of the Root Symmetry mode x3 , where I clicked 12 times (that's 3^11 + 3^10 + 3^9 + ... + 3^0 parts)
(And that is as much parts as I can place if I don't want to wait minutes for next iteration to spawn)

There might be some bugs concerning part destruction and replacing new ones after that.

UknownPI committed:
Symmetry

Controls are :
U I -> Change Iteration Count
H J -> Change Symmetry Count
+ Added Some 3d models

dagriefaa commented on Assets/Code/VehicleEditor/EditSystem.cs:
```diff
@@ -16,6 +21,9 @@
 /// </summary>
 partial class EditSystem : SystemBase {
     private const float SNAP_DISTANCE = .5f;
+    int lastID = 1; // Each time parts are placed (dont matter how many, if it is just in 1 click) they all have the same ID
```
systems shouldn't have state - add variables to EditSystemData

dagriefaa commented on Assets/Code/VehicleEditor/EditSystem.cs:
```diff
@@ -105,8 +128,36 @@ partial class EditSystem : SystemBase {
+                        */
+                    RootSymmetryMode(actionQueue: actionQueue, part: partPrefab, parentPart: parentPart, symmetryCount: sym, placeTransform: placeTransform,MaxIt:it);
+                    /*
+                    var symmetryCount = SystemAPI.GetSingleton<EditSystemData>().SymmetryCount;
```
don't leave commented code, just rip it out - if we want it back we can always check the commit history

UknownPI replied:
ok

dagriefaa commented on Assets/Code/VehicleEditor/EditSystem.cs:
```diff
@@ -307,4 +375,168 @@ partial class EditSystem : SystemBase {
+            Debug.Log("");
+            int iteration = SystemAPI.GetComponent<Part>(parentPart).layer; // ok idk if iteration is the right term but it will do just fine for now
+
+            int[] symmetryCounts = new int[iteration];
```
should use data structures from `Unity.Collections` (e.g. `Unity.Collections.NativeArray<T>` (i think?)) - not that important for this use case or this prototype but reduces garbage collection overhead

UknownPI replied:
Not sure how to do that

dagriefaa replied:
docs here https://docs.unity3d.com/Packages/com.unity.collections@2.4/manual/index.html

there are examples of `NativeList` usage in this class e.g. line 226

dagriefaa commented on Assets/Code/VehicleEditor/EditSystem.cs:
```diff
@@ -307,4 +375,168 @@ partial class EditSystem : SystemBase {
+    // dont know if not using ref will be an issue
+    // THERE MIGHT BE AN ISSUE , IF A PART IS REGISTERED AS ANOTHER PART SIBLING BUT THEN THIS FIRST PART IS DESTROYED
+    // IF A NEW PART IS PLACED WITH THE SAME VALUE AS THE DESTROYED ONE THERE MIGHT BE SOME ISSUE -> to fix this the deleted part's reference should be cleaned from his siblings and parents
+    void RootSymmetryMode(EntityCommandBuffer actionQueue,Entity part,Entity parentPart,int symmetryCount,LocalTransform placeTransform,int MaxIt) {
```
should remove unused arguments here

(note that EntityCommandBuffer is a struct, which means it's passed by value like a number, which means you have to either use `ref` or return it if you make changes to it in this method as otherwise you'll lose the changes)

UknownPI replied:
All Arguments are used here

UknownPI replied:
Also if I add ref here I get an error later line 483 , when I pass that entity command buffer to another function (rotatingpartplacement) 

`Error	CS1628	Cannot use ref, out, or in parameter 'actionQueue' inside an anonymous method, lambda expression, query expression, or local function`

dagriefaa replied:
you need to pass it as a `ref` parameter on `RecursifFor` as well, because you can't hand a `ref` parameter directly to the inside of an inner function

probably better to not make `RecursifFor` an inner function at all and put it in the class scope?

dagriefaa commented on Assets/Code/VehicleEditor/EditSystem.cs:
```diff
@@ -307,4 +375,168 @@ partial class EditSystem : SystemBase {
+    }
+}
+
+public struct Parente : IComponentData
```
put component structs in their own files (or, better yet, put their functions in existing component structs)

UknownPI replied:
ok

dagriefaa commented on Assets/Code/VehicleEditor/EditSystem.cs:
```diff
@@ -307,4 +375,168 @@ partial class EditSystem : SystemBase {
+
+        var currentPart = parentPart;
+        for (int i = 0; i < iteration; i++)
+        {
```
maybe you could use a list of structs instead of multiple arrays?

UknownPI replied:
I'm not sure to understand , could you write an example

dagriefaa replied:
```cs
struct Placement {
    public int symmetryCounts;
    public int id;
    public int selfplace;
    public Entity entity;
}

Placement[] placements = new Placement[iteration];

var currentPart = parentPart;
for (int i = iteration - 1; i >= 0; i--)
{
    placements[i] = new SomeStruct {
        symmetryCounts = EntityManager.GetBuffer<Siblings>(currentPart).Length,
        id = SystemAPI.GetComponent<Part>(currentPart).id,
        entity = currentPart,
        selfplace = SystemAPI.GetComponent<Part>(currentPart).selfplace
    };
    if (i > 0) {currentPart = SystemAPI.GetComponent<Parente>(currentPart).Value; } // root has no parent
}
// etc etc
```
probably not 100% correct but you get the idea

dagriefaa commented on Assets/Code/VehicleEditor/EditSystem.cs:
```diff
@@ -141,17 +192,34 @@ partial class EditSystem : SystemBase {
-
         actionQueue.SetComponent(newPart, placeTransform);
+        
+        /*if (parentPart != Entity.Null) // this is not needed since there is no more parent (seperated into to methodes) (might wanna merge them back into one in the future)
```
does deleting part trees still work?

UknownPI replied:
Deleting Part Tree still works but it might need some changes :
- The part which is deleted should be erased from his siblings to Entity.Null
- And I'm not sure if you already did this but it might be good to delete that part from the buffer of children of the parent of deleted part

dagriefaa reviewed (commented):
@UnknownDude1 see comments - haven't tested this stuff in-depth, so just providing surface-level feedback

UknownPI committed:
Some Changes suggested by dagriefaa

+ "don't leave commented code, just rip it out "
+ "systems shouldn't have state - add variables to EditSystemData"
+ "put component structs in their own files"
